crates/snapbox/src/data.rs

```diff
@@ -96,34 +122,47 @@ impl Data {
         match &self.inner {
             DataInner::Binary(_) => None,
             DataInner::Text(data) => Some(data.to_owned()),
+            #[cfg(feature = "json")]
+            DataInner::Json(value) => Some(serde_json::to_string_pretty(value).unwrap()),
         }
     }
 
     pub fn to_bytes(&self) -> Vec<u8> {
         match &self.inner {
             DataInner::Binary(data) => data.clone(),
             DataInner::Text(data) => data.clone().into_bytes(),
+            #[cfg(feature = "json")]
+            DataInner::Json(value) => serde_json::to_vec_pretty(value).unwrap(),
         }
     }
 
     pub fn try_coerce(self, format: DataFormat) -> Self {
-        match format {
-            DataFormat::Binary => Self::binary(self.to_bytes()),
-            DataFormat::Text => match self.inner {
-                DataInner::Binary(data) => {
-                    if is_binary(&data) {
-                        Self::binary(data)
-                    } else {
-                        match String::from_utf8(data) {
-                            Ok(data) => Self::text(data),
-                            Err(err) => {
-                                let data = err.into_bytes();
-                                Self::binary(data)
-                            }
-                        }
-                    }
+        match (self.inner, format) {
+            (DataInner::Binary(inner), DataFormat::Binary) => Self::binary(inner),
+            (DataInner::Text(inner), DataFormat::Text) => Self::text(inner),
+            #[cfg(feature = "json")]
+            (DataInner::Json(inner), DataFormat::Json) => Self::json(inner),
+            (DataInner::Binary(inner), _) => {
+                if is_binary(&inner) {
+                    Self::binary(inner)
+                } else if let Ok(str) = String::from_utf8(inner.clone()) {
+                    Self::text(str).try_coerce(format)
+                } else {
+                    Self::binary(inner)
+                }
+            }
+            (DataInner::Text(inner), DataFormat::Binary) => Self::binary(inner.into_bytes()),
+            #[cfg(feature = "json")]
+            (DataInner::Text(inner), DataFormat::Json) => {
+                match serde_json::from_str::<serde_json::Value>(&inner) {
+                    Ok(json) => Self::json(json),
+                    Err(_) => Self::text(inner),
                 }
-                DataInner::Text(data) => Self::text(data),
+            }
+            #[cfg(feature = "json")]
+            (DataInner::Json(inner), _) => match serde_json::to_vec_pretty(&inner) {
+                Ok(bin) => Self::binary(bin).try_coerce(format),
+                Err(_) => Self::json(inner),
```

Comment:
Let's use `to_string_pretty` instead to avoid extra `is_binary` checks and utf8 validation since json should always be utf8